### ISSUE TITLE
Add a realistic e2e SCTP test

### ISSUE THREAD

**Issue by aojea** [kubernetes/kubernetes#103150]
#### What would you like to be added:

Add an e2e test with a real SCTP scenario.

#### Why is this needed:

We are only testing basic SCTP functionality, i.e. my SCTP packet reaches the corresponding endpoint, however that has demonstrated to be insufficient.

We need to have and e2e test with a real SCTP scenario so we can be sure that kubernetes meets the requirements and the expectations of SCTP users.

/sig network

**Comment by aojea** [comment]
quoting @ravens 

>  was thinking we could use srsRAN as a realistic demo payload. They provide a self contained EPC, a eNB and UE that can run with a fake RF mode ? Here is the EPC https://gist.github.com/ravens/a8a151c7cb219e822b37ab512a33ca41, and config file can be tuned in an Initcontainer.

This seems a good fit

**Comment by uablrek** [comment]
The payload is not really significant for tests except for expected packet sizes (fragmentation/PMTU for IPv4 and IPv6?). What has proven problematic is dynamics in fail-over/restart cases, and of course multihoming. It is _possible_ the `nmap-ncat` is sufficient for e2e test, but even to get that accepted seems to be a struggle https://github.com/kubernetes/kubernetes/pull/94597.

**Comment by ravens** [comment]
@uablrek then maybe using something based on [pycrate](https://github.com/P1sec/pycrate) to fake different state of SCTP connection/payload size. 
One question, regarding multihoming I guess the idea is to test with Multus ?

**Comment by uablrek** [comment]
> I guess the idea is to test with Multus

Not really no. Multus is out of scope for k8s e2e (I think). But it seems some sctp users on K8s using multihoming and a sctp service. I have currently no idea how. The node must have 2 interfaces (at least) and if the client connect through the service the HB may in theory be sent by the server (pod) via the second interface. However I can't figure how the NAT problem is resolved.

**Comment by uablrek** [comment]
### Who uses the SCTP support in K8s?

IMHO this question must be answered before we can define any realistic e2e K8s test for sctp.

As a telco user I can't see _any_ use for sctp without multihoming. To combine multihoming with load-balancing and NAT (used in K8s services) seem impossible in a good way ([rfc3257](https://datatracker.ietf.org/doc/html/rfc3257#section-4)). It seems multus, possible with sriov/dpdk, or to run in host namespace ("hostNetwork:true") are the only options. But that is done without K8s support.

The only user of non-multihoming sctp I could find is [WebRTC](https://en.wikipedia.org/wiki/WebRTC). But first it seems like an old thing, and second it uses sctp-over-udp.

**Comment by ravens** [comment]
@uablrek there are use of SCTP without multihoming ink k8s, i.e. for example the S1AP/NGAP interface for eNB/gNB where multihoming is not used, MME polling is rather used. In general SCTP is a transport protocol and multi homing is not mandatory ?

**Comment by uablrek** [comment]
Ok, then it should be documented that K8s does not support sctp multihoming and as such it does not have to be tested in K8s e2e tests.

I can't see any need to emulate a RAN or use pycrate, same as we don't setup a fake web-shop to test tcp. It will only make e2e tests hard to maintain IMHO. Basically the same level of e2e tests done for tcp/udp should be sufficient, and I think `nmap-ncat` can be used for that.

**Comment by aojea** [comment]
> same as we don't setup a fake web-shop to test tcp

we do install a webshop to test e2e, https://github.com/kubernetes/kubernetes/tree/master/test/e2e/testing-manifests/guestbook

> I can't see any need to emulate a RAN or use pycrate,

I have to disagree here, having only e2e test that exercise happy scenarios "create service - ping service" are not providing any signal on quallity, and I will categorize those more close to integration than e2e

> Basically the same level of e2e tests done for tcp/udp should be sufficient,

this is a common misunderstanding in Kubernetes, everything in the e2e framework exercises tcp and udp, despite there are some specifics tests only for those protocols,

E2E has to be about real scenarios, so the project knows that when an user installs or deploy something it will work, becuse we tested it. Traditionally this was done by QE/QA departments, but now everything can be automated and MUST be automated

**Comment by uablrek** [comment]
> we do install a webshop to test e2e, https://github.com/kubernetes/kubernetes/tree/master/test/e2e/testing-manifests/guestbook

Oh, cool. I stand corrected. @aojea You are the king of K8s e2e :smile: 

I will investigate how to use sctp in K8s the next couple of months, hence my interrest. A legit question is "Can we just use the K8s sctp?". In my case the answer is "No, we need solid multihoming". I hoped that is was possible, mostly based on https://github.com/kubernetes/kubernetes/issues/102128. But even if it _may_ be possible to some extent it seems to be a cludge and fragile.

**Comment by uablrek** [comment]
This comment should be re-phrased; https://kubernetes.io/docs/concepts/services-networking/service/#caveat-sctp-multihomed

It makes you think K8s actually has some support for multihoming. It would be better to say something like;

> Multihoming is not supported by the K8s sctp service.

Even with a "CNI plugin can support the assignment of multiple interfaces and IP addresses to a Pod" (e.g multus) the current kube-proxy can't make use of them. Load-balancing+multihoming is a _really_ difficult combination.

Also, does anybody know what;

> NAT for multihomed SCTP associations requires special logic in the corresponding kernel modules

referes to? I have searched for this but not found anything in Linux (but I havn't scanned the kernel source). From [rfc3257](https://datatracker.ietf.org/doc/html/rfc3257#section-4);

> the NAT can have internal Application Layer Gateway (ALG) which will intelligently translate the IP addresses in the INIT and INIT ACK chunks.

This sounds rather horrible to me, especially since sctp has a checksum that must be re-computed.

**Comment by ravens** [comment]
Probably a mention of that multihoming is requiring either 1) specific use of CNI solution for multiple IP allocation (multus, calico enterprise etc...) and 2) alternate kube-proxy approach for compatibility with NAT for Service IP in that particular use case. Nevertheless I would not reuse SCTP usage to multihoming, @uablrek.

**Comment by ravens** [comment]
@aojea based on your guestbook example, perhaps a 5G example could be proposed - i.e. [UERANSIM](https://github.com/aligungr/UERANSIM) with either [open5gs](https://github.com/open5gs)/[free5gs](https://github.com/free5gc/free5gc), but then we can only test the control plane (NGAP) as the user plan (GTP, UDP based) is requiring a GTP kernel driver for both project. 

Perhaps a subset of the original script from [corenet](https://github.com/mitshell/corenet) (pycrate based) is more convenient ? As we could build an example only for the control plane (NGAP in 5G or S1AP in 4G, which both use SCTP as transport). A minimal working Dockerfile for using pycrate is provided [here]. (https://gist.github.com/ravens/6bc15c339d6ae58042e63f332ccab5ca)

**Comment by ravens** [comment]
A good candidate for a problematic packet S1AP/NGAP exchange could be the the ["UE capability Information"](https://www.sharetechnote.com/html/Handbook_LTE_UE_Capability.html), typically at some point the terminal send a message to the core in 4G/5G detailing the actual capability (i.e. in terms of radio, VoLTE support etc) to the core. This packet sent over SCTP ends up being quite large (>3k bytes) and is typically a good indicator of MTU problems.
